### PR TITLE
Rsu health monitor

### PR DESCRIPTION
**Pull request by dan-du-car**
# PR Details
## Description
RSU health monitor plugin is a V2xHub plugin that interface with its connected RSU directly via SNMP protocol. The plugin is responsible for monitoring the connected RSU status by constantly ping RSU device. For detailed design of this plugin, refer to https://usdot-carma.atlassian.net/wiki/spaces/WFD2/pages/2640740360/RSU+Health+Monitor+Plugin+Design .

## Related Issue
NA

## Motivation and Context
Data visualization

## How Has This Been Tested?
Integration test

## Types of changes

- [ ] Defect fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that cause existing functionality to change)

## Checklist:

- [ ] I have added any new packages to the sonar-scanner.properties file
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
[V2XHUB Contributing Guide](https://github.com/usdot-fhwa-OPS/V2X-Hub/blob/develop/Contributing.md) 
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/tmx/Messages/include/RSUStatusMessage.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+#include "MessageTypes.h"
+
+
+namespace tmx::messages {
+
+
+class RSUStatusMessage : public tmx::message
+{
+	public:
+		RSUStatusMessage() {}
```
Comment (paulbourelly999) [inline src/tmx/Messages/include/RSUStatusMessage.h:14]: Does this message not have any attributes?

File: src/tmx/TmxUtils/src/SNMPClient.cpp [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+    // Client defaults to SNMPv3
+    snmp_client::snmp_client(const std::string& ip, const int& port, const std::string& community, 
+        const std::string &snmp_user, const std::string &securityLevel, const std::string &authPassPhrase, int snmp_version, int timeout)
+
+        : ip_(ip), port_(port), community_(community),snmp_version_(snmp_version), timeout_(timeout)
+    {
+        
+        PLOG(logDEBUG1) << "Starting SNMP Client";
+        PLOG(logDEBUG1) << "Target device IP address: " << ip_;
+        PLOG(logINFO) << "Target device NTCIP port: " << port_;
```
Comment (paulbourelly999) [inline src/tmx/TmxUtils/src/SNMPClient.cpp:14]: This logging does not may sense to me. NTCIP port is probably incorrect here. Also we probably want all these log statements to share a log level.

File: src/v2i-hub/RSUHealthMonitorPlugin/manifest.json [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+		{
+            "key":"RSUOIDConfigMap",
+            "default":"{\"RSUOIDConfig\":[{\"RsuField\":\"rsuGpsOutputString\",\"OID\":\"iso.0.15628.4.1.8.5.0\",\"Required\":false},{\"RsuField\":\"rsuID\",\"OID\":\"iso.0.15628.4.1.17.4.0\",\"Required\":false},{\"RsuField\":\"rsuMibVersion\",\"OID\":\"iso.0.15628.4.1.17.1.0\",\"Required\":false},{\"RsuField\":\"rsuFirmwareVersion\",\"OID\":\"iso.0.15628.4.1.17.2.0 \",\"Required\":false},{\"RsuField\":\"rsuManufacturer\",\"OID\":\"iso.0.15628.4.1.17.5.0\",\"Required\":false},{\"RsuField\":\"rsuIFMIndex\",\"OID\":\"iso.3.6.1.2.1.1.7.0\",\"Required\":false},{\"RsuField\":\"rsuIFMPsid\",\"OID\":\"iso.3.6.1.2.1.1.7.0\",\"Required\":false},{\"RsuField\":\"rsuIFMDsrcMsgId\",\"OID\":\"iso.3.6.1.2.1.1.7.0\",\"Required\":false},{\"RsuField\":\"rsuIFMTxMode\",\"OID\":\"iso.3.6.1.2.1.1.7.0\",\"Required\":false},{\"RsuField\":\"rsuIFMTxChannel\",\"OID\":\"iso.3.6.1.2.1.1.7.0\",\"Required\":false},{\"RsuField\":\"rsuIFMEnable\",\"OID\":\"iso.3.6.1.2.1.1.7.0\",\"Required\":false},{\"RsuField\":\"rsuIFMStatus\",\"OID\":\"iso.3.6.1.2.1.1.7.0\",\"Required\":false},{\"RsuField\":\"rsuMode\",\"OID\":\"iso.0.15628.4.1.99.0\",\"Required\":false}]}",
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/manifest.json:42]: This does not make sense to me. Aren't these OIDs standardized. Maybe it would be better to specify a document version here and then have different headers that contain the actual OIDs.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+                        Json::FastWriter fasterWirter;
+                        string json_str = fasterWirter.write(rsuStatusJson);
+                        tmx::messages::RSUStatusMessage sendRsuStatusMsg;
+                        sendRsuStatusMsg.set_contents(json_str);
+                        BroadcastMessage(sendRsuStatusMsg);
+                        PLOG(logINFO) << "Broadcast RSU status:  " << json_str;
+                    }
+                }
+            }
+            catch (std::exception &ex)
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:102]: This should be const

File: src/tmx/Messages/include/RSUStatusMessage.h (hunk 1 above)
Comment (dan-du-car) [reply src/tmx/Messages/include/RSUStatusMessage.h:14]: No, we just need the content which is from tmx::message.

File: src/tmx/TmxUtils/src/SNMPClient.cpp (hunk 2 above)
Comment (dan-du-car) [reply src/tmx/TmxUtils/src/SNMPClient.cpp]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/c0bac101e0cef01c154929024be6ab01b263785b

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp (hunk 4 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/c0bac101e0cef01c154929024be6ab01b263785b

File: src/tmx/TmxUtils/src/RSU_MIB_4_1.h [hunk 5]
```diff
@@ -0,0 +1,69 @@
+#pragma once
+namespace tmx
```
Comment (paulbourelly999) [inline src/tmx/TmxUtils/src/RSU_MIB_4_1.h:2]: You can do something like this to make this less verbose
```
namespace tmx::utils::rsu41::mid::oid
```

File: src/v2i-hub/RSUHealthMonitorPlugin/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+#############
+enable_testing()
+include_directories(${PROJECT_SOURCE_DIR}/src)
+add_library(${PROJECT_NAME}_lib src/RSUHealthMonitorWorker.cpp)
+target_link_libraries(${PROJECT_NAME}_lib PUBLIC ${TMXAPI_LIBRARIES} 
+                                        ${ASN_J2735_LIBRARIES} 
+                                        ${MYSQL_LIBRARIES} 
+                                        ${MYSQLCPPCONN_LIBRARIES}
+                                        tmxutils
+                                        NemaTode
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/CMakeLists.txt:26]: What is nematode?

File: src/v2i-hub/RSUHealthMonitorPlugin/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+                                        ${UUID_LIBRARY})
+set(BINARY ${PROJECT_NAME}_test)
+file(GLOB_RECURSE TEST_SOURCES LIST_DIRECTORIES false test/*.h test/*.cpp)
+set(SOURCES ${TEST_SOURCES} WORKING_DIRECTORY ${PROJECT_SOURCE_DIR}/test)
+add_executable(${BINARY} ${TEST_SOURCES})
+add_test(NAME ${BINARY} COMMAND ${BINARY})
+target_link_libraries(${BINARY} PUBLIC ${PROJECT_NAME}_lib gtest NemaTode)
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/CMakeLists.txt:33]: you do not need to link NemaTode here if you made it a public depedency of you lib and you are including you lib

File: src/v2i-hub/RSUHealthMonitorPlugin/manifest.json [hunk 8]
```diff
@@ -0,0 +1,51 @@
+{
+    "name": "RSUHealthMonitor",
+    "description": "Monitor RSU health status",
+    "version": "@PROJECT_VERSION@",
+    "exeLocation": "/bin/RSUHealthMonitorPlugin",
+    "coreIpAddr":"127.0.0.1",
+	"corePort":24601,
+    "messageTypes": [],
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/manifest.json:8]: You should probably add the RSU Health message to message types here.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+
+    void RSUHealthMonitorPlugin::OnConfigChanged(const char *key, const char *value)
+    {
+        PluginClient::OnConfigChanged(key, value);
+        UpdateConfigSettings();
+    }
+
+    void RSUHealthMonitorPlugin::PeriodicRSUStatusReq()
+    {
+        while (true)
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:52]: Not a fan of non exitable loops. Consider using one of the built in tmx wrappers on thread instead of thread directly like `ThreadTimer` or `ThreadWorker`. Even if you continue to use thread directly there should probably be some sort of exit condition. If you extend thread working, you can implement the DoWork function and it makes this logic unit testable since it will no longer be inside you plugin class.

File: src/v2i-hub/RSUHealthMonitorPlugin/CMakeLists.txt (hunk 6 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/CMakeLists.txt:26]: https://github.com/ckgt/NemaTode We use this to parse the RSU GPS Nmea string.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+            }
+            catch (const std::exception &ex)
+            {
+                PLOG(logERROR) << ex.what();
+            }
+            this_thread::sleep_for(chrono::seconds(_interval));
+        }
+    }
+
+    Json::Value RSUHealthMonitorPlugin::getRSUStatus()
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:87]: Is there are reason this is not in the RSU Worker. Here it is not unit testable?

File: src/v2i-hub/RSUHealthMonitorPlugin/manifest.json (hunk 8 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/manifest.json:8]: What is the purpose of adding this message types here?

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    {
+        if (!_rsuWorker)
+        {
+            PLOG(logERROR) << "RSU status update call failed due to fail to initialize RSU worker!";
+            return Json::nullValue;
```
[the remark this answers is not in the document]
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:92]: The caller checks null value to determine the action rather than requiring all callers to catch them later.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp (hunk 10 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:87]: It is sending the SNMP request which requires the connection to the RSU, and also require the global RSU Status config variable from the plugin.

**Review by paulbourelly999 — CHANGES_REQUESTED** [review]
Please check out the sonar scan code smells. A lot of these seem pretty easy to address. Also you need to update the sonar.properties file to include your new plugin

**Comment by dan-du-car** [comment]
> Please check out the sonar scan code smells. A lot of these seem pretty easy to address. Also you need to update the sonar.properties file to include your new plugin

Added https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/5f7858e46c999a341f91ed0f8685ca8b14711552

File: src/v2i-hub/RSUHealthMonitorPlugin/CMakeLists.txt (hunk 7 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/CMakeLists.txt]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/5f7858e46c999a341f91ed0f8685ca8b14711552

File: src/tmx/TmxUtils/src/RSU_MIB_4_1.h (hunk 5 above)
Comment (dan-du-car) [reply src/tmx/TmxUtils/src/RSU_MIB_4_1.h]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/d46893578a0d09dfdf8decc8166c102da3e3a449

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp (hunk 9 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:52]: Is there any issue with retrieving the RSU info and broadcast immediately? I am trying to avoid the complexity of two threads accessing the same global variable where the RSU status is stored.

Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/7a4238acb369a0a8baa26b437c9b0ab59e48960a

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp (hunk 10 above)
Comment (paulbourelly999) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:90]: Both the SNMP Client and any configuration can be passed to another class to make this call right? Just my opinion, but when creating new V2X-Hub Plugins since we can not unit test the code there, we should be moving most of it to other cpp files. Also for maintainability reasons we should not couple the logic for making snmp request, json serialization and the plugin core logic together if we do not need to.

File: src/v2i-hub/RSUHealthMonitorPlugin/manifest.json (hunk 8 above)
Comment (paulbourelly999) [reply src/v2i-hub/RSUHealthMonitorPlugin/manifest.json:8]: I guess in this case it won't work but for plugins that receive a certain message, it adds a status item that includes a count for the number of that type of message received by the plugin.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        // Send SNMP call to RSU periodically at configurable interval.
+        _timerThId = _rsuStatusTimer->AddPeriodicTick([this]()
+                                                      {
+            // Periodic SNMP call to get RSU status based on RSU MIB version 4.1
+            auto rsuStatusJson =  _rsuWorker->getRSUStatus(_rsuMibVersion, _rsuIp, _snmpPort, _securityUser, _authPassPhrase, _securityLevel, SEC_TO_MICRO);
+            PLOG(logINFO) << "Updating _interval: " << _interval;
+            //Broadcast RSU status periodically at _interval
+            BroadcastRSUStatus(rsuStatusJson); },
+                                                      std::chrono::milliseconds(_interval * SEC_TO_MILLI));
+        _rsuStatusTimer->Start();
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:24]: Can we move this start to when the Plugin is enabled on config changes.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+            auto latitude_str = ss.str();
+            std::stringstream sss;
+            sss << std::setprecision(8) << std::fixed << gps.fix.longitude << std::endl;
+            auto longitude_str = sss.str();
+            result.insert({std::stod(latitude_str), std::stod(longitude_str)});
+            PLOG(logDEBUG) << "Parse GPS NMEA string: " << gps_nmea_data << ". Result (Latitude, Longitude): (" << latitude_str << "," << longitude_str << ")";
+        }
+        catch (nmea::NMEAParseError &e)
+        {
+            fprintf(stderr, "Error:%s\n", e.message.c_str());
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp:112]: Why are we using fprintf instead of PLOG

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+            std::stringstream ss;
+            ss << std::setprecision(8) << std::fixed << gps.fix.latitude << std::endl;
+            auto latitude_str = ss.str();
+            std::stringstream sss;
+            sss << std::setprecision(8) << std::fixed << gps.fix.longitude << std::endl;
+            auto longitude_str = sss.str();
+            result.insert({std::stod(latitude_str), std::stod(longitude_str)});
+            PLOG(logDEBUG) << "Parse GPS NMEA string: " << gps_nmea_data << ". Result (Latitude, Longitude): (" << latitude_str << "," << longitude_str << ")";
+        }
+        catch (nmea::NMEAParseError &e)
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp:110]: Should be constant

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+        return result;
+    }
+
+    Json::Value RSUHealthMonitorWorker::getRSUStatus(const RSUMibVersion &mibVersion, const string &_rsuIp, uint16_t &_snmpPort, const string &_securityUser, const string &_authPassPhrase, const string &_securityLevel, long timeout)
+    {
+        auto rsuStatusConfigTbl = GetRSUStatusConfig(mibVersion);
+        if (rsuStatusConfigTbl.size() == 0)
+        {
+            PLOG(logERROR) << "RSU status update call failed due to the RSU status config table is empty!";
+            return Json::nullValue;
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp:123]: Why are we printing error and returning null instead of throwing an exception

File: src/tmx/TmxUtils/src/SNMPClient.cpp [hunk 16]
```diff
@@ -0,0 +1,306 @@
...
+        return true;
+    }
+
+    // Backup GET function for use with prexisting ERVCloudForwarding client
+    std::string snmp_client::SNMPGet(const std::string &req_oid)
+    {
+        snmp_pdu *response;
+
+        std::string result = "";
+        auto pdu = snmp_pdu_create(SNMP_MSG_GET);
```
Comment (paulbourelly999) [inline src/tmx/TmxUtils/src/SNMPClient.cpp:241]: There is a member variable with the same name and overlapping scope. We should probably reconsider naming here

File: src/tmx/TmxUtils/CMakeLists.txt [hunk 17]
```diff
@@ -1,7 +1,9 @@
 PROJECT ( tmxutils CXX )
 FILE (GLOB_RECURSE HEADERS "src/" "*.h*")
 FILE (GLOB_RECURSE SOURCES  "src/" "*.c*")
-
```
Comment (paulbourelly999) [inline src/tmx/TmxUtils/CMakeLists.txt:4]: This is optional but I think it would be helpful if we separate some of this into sub packages like `TmxUtils/src/snmp_client` similar to what we have done with kafka.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+    struct RSUFieldOIDStruct
+    {
+        string field;
+        string oid;
+        bool required; // Indicate whether this field is required to before broadcasting the RSU status.
+    };
+
+    /**
+     * RSUStatusTable is custome defined RSU status information.
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h:35]: *custom

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+     * The fields are the subset of fields from RSU MIB definition https://github.com/certificationoperatingcouncil/COC_TestSpecs/blob/master/AppNotes/RSU/RSU-MIB.txt
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h:36]: The fields are a subset of the fields from the RSU MIB definition used to quantify the health of the RSU.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+     */
+    using RSUStatusConfigTable = vector<RSUFieldOIDStruct>;
+
+    class RSUHealthMonitorWorker
+    {
+    private:
+        // A map of RSU MIB version used and RSUStatusTable
+        shared_ptr<map<RSUMibVersion, RSUStatusConfigTable>> _RSUSTATUSConfigMapPtr;
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h:44]: Why do we have a map between mib version and RSU Status Config table? Will we populate this map with all versions regardless of the configured version and then just use the configured version?

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+        RSUStatusConfigTable GetRSUStatusConfig(const RSUMibVersion &mibVersion) const;
+
+        /**
+         * @brief determine if all required fields in the RSU config map _RSUSTATUSConfigMapPtr present in the input fields
+         * Use _RSUSTATUSConfigMapPtr RSU status config map that defines all fields and whether the fields are required.
+         * @param RSUMibVersion RSU MIB version
+         * @param vector<string> Input RSU fields to verify
+         * @return True if all required fields found. Otherwise, false.
+         */
+        bool isAllRequiredFieldsPresent(const RSUMibVersion &mibVersion, const vector<string> &fields) const;
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h:68]: Can we rename this method to validate. Also I am not sure I understand this method. Is it meant to validate a RSU status message or is it meant to confirm all the required OIDs are obtainable from SNMP client?

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+
+        /**
+         * @brief Parse NMEA GPS sentense and return GPS related data
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h:71]: *Sentence Although maybe we want to reword this to be Parse NMEA GPS String ...
Also what is GPS related data? Is this just a long lat coordinate. Why do we need a map if its just a long lat coordinate?

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp [hunk 23]
```diff
@@ -0,0 +1,175 @@
...
+            RSUFieldOIDStruct rsuMode = {"rsuMode", RSU_MODE, true};
+            rsuStatusTbl.push_back(rsuMode);
+
+            RSUFieldOIDStruct rsuChanStatus = {"rsuChanStatus", RSU_CHAN_STATUS, true};
+            rsuStatusTbl.push_back(rsuChanStatus);
+        }
+        return rsuStatusTbl;
+    }
+
+    bool RSUHealthMonitorWorker::isAllRequiredFieldsPresent(const RSUMibVersion &mibVersion, const vector<string> &fields) const
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp:65]: If this is a validation method I suggest for clarity sake you pass in the config directly instead of passing in the mib version and calling Get config. I think it is cleaner for the validate method to take parameters ( message to validate, list of required fields)

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+    {
+        _rsuWorker = std::make_shared<RSUHealthMonitorWorker>();
+        _rsuStatusTimer = make_unique<ThreadTimer>();
+        UpdateConfigSettings();
+
+        // Send SNMP call to RSU periodically at configurable interval.
+        _timerThId = _rsuStatusTimer->AddPeriodicTick([this]()
+                                                      {
+            // Periodic SNMP call to get RSU status based on RSU MIB version 4.1
+            auto rsuStatusJson =  _rsuWorker->getRSUStatus(_rsuMibVersion, _rsuIp, _snmpPort, _securityUser, _authPassPhrase, _securityLevel, SEC_TO_MICRO);
```
Comment (paulbourelly999) [inline src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:19]: This is confusing to me. So you get the RSU status here by querying the SNMP server and you return a JSON. Then you wait until the broadcast method to check that the required information is present? You do not have to change this, it just seems weird to me. I would imagine that the method that gets the RSU status either directly turns it into the tmx message (including validation) or the next method turns it into a tmx message.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp (hunk 15 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp:123]: We encourage users to update the RSU config table by updating the mib version with the supported version, but it is not a fatal program issue.

File: src/tmx/TmxUtils/CMakeLists.txt (hunk 17 above)
Comment (dan-du-car) [reply src/tmx/TmxUtils/CMakeLists.txt:4]: I would leave this to enhancement later if we would like to reorganize all the files.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h (hunk 20 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h:44]: yes, we will populate the map with all versions if exist like NTCIP1218 and RSU4.1. But the users can choose which version to use by specifying the mib version through the UI?

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h (hunk 21 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h:68]: sure, update to validate. Yes, it is to make sure we obtain all the required fields.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h (hunk 22 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h:71]: Yes, what data structure do you suggest if not map?

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp (hunk 24 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp:19]: If the JSON is empty due to some SNMP timeout or other issue, we stop any further logic like creating TMX message, broadcasting message.

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp (hunk 23 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/a49366ea8030afa8c668ec5879cf8e2d5cf69065

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h (hunk 21 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/a49366ea8030afa8c668ec5879cf8e2d5cf69065

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h (hunk 19 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/a49366ea8030afa8c668ec5879cf8e2d5cf69065

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h (hunk 18 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.h]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/a49366ea8030afa8c668ec5879cf8e2d5cf69065

File: src/tmx/TmxUtils/src/SNMPClient.cpp (hunk 16 above)
Comment (dan-du-car) [reply src/tmx/TmxUtils/src/SNMPClient.cpp]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/a49366ea8030afa8c668ec5879cf8e2d5cf69065

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp (hunk 14 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/a49366ea8030afa8c668ec5879cf8e2d5cf69065

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp (hunk 13 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/a49366ea8030afa8c668ec5879cf8e2d5cf69065

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp (hunk 12 above)
Comment (dan-du-car) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorPlugin.cpp]: Addressed https://github.com/usdot-fhwa-OPS/V2X-Hub/pull/564/commits/a49366ea8030afa8c668ec5879cf8e2d5cf69065

File: src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp (hunk 15 above)
Comment (paulbourelly999) [reply src/v2i-hub/RSUHealthMonitorPlugin/src/RSUHealthMonitorWorker.cpp:123]: Exceptions do not need to be fatal but this is fine for now.

**Review by paulbourelly999 — APPROVED** [review]
